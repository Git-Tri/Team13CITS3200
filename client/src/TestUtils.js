const trueErrorLog = console.error;

/**
 * okay this is a weird one to explain
 * basically there an issue with our dev setup (web-pack/react/create react app)
 * where if you test when a component fails it will spam the error log with
 * a message telling you the component failed 
 * since we are testing that, this is completely unneccessary
 * as such i am filtering the console.error to exclude these messages 
 * to give a more readable test output
 * using monkey patching: https://davidwalsh.name/monkey-patching
 */
export function filterConsoleError()
{

    
    console.error = (e) => 
    {

        
        if(typeof(e) === "string" && ! e.includes("React will try to recreate this component tree from scratch using the error boundary you provided"))
        {

            trueErrorLog(e);
        
        }
        else if(typeof(e) !== "string")
        {

            trueErrorLog(e);

        }
        
    }

}

export function unfilterConsoleError()
{

    console.error = trueErrorLog;

}