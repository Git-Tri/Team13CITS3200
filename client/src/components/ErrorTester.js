import React from "react";

/**
 * Catches an error and uses the onError callback to allow tests to pass
 * this component is meant for testing purposes only
 * when testing place the error component as a child to this component 
 * then pass in an onError prop or check the state of this component
 * derived from: https://codepen.io/gaearon/pen/wqvxGa?editors=0010
 */
class ErrorTester extends React.Component {
    constructor(props) {
      super(props);
      this.state = { error: null, errorInfo: null };
    }
    
    componentDidCatch(error, errorInfo) {
      this.setState({
        error: error,
        errorInfo: errorInfo
      })

      this.props.onError(error,errorInfo);
    }
    
    render() {
      if (this.state.errorInfo) {
        return (
          <p> Good news an error occurred</p>
          );
      }
      return this.props.children;
    }  
  }

  export default ErrorTester;
