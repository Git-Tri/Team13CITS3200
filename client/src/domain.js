/**
 * A class representing unstructured data
 * for list you may want to leave the data field undefined
 * NOTE: this code is duplicated on client and server so is only tested on the server
 */
export class UnstructuredData
{

    constructor(id,matchid,title,author,url,published,extracted,data)
    {

        this.id = id;
        this.matchid = matchid;
        this.title = title;
        this.author = author;
        this.url = url;
        this.published = published;
        this.extracted = extracted;
        this.data = data; 

    }

}

/**
 * A class representing structured data, for lists you may want to leave data undefined
 */
export class StructuredData
{

    constructor(id,date,home,away,competitionID,competitionName,data)
    {

        this.id = id;
        this.date = date;
        this.home = home;
        this.away = away;
        this.competitionID = competitionID;
        this.competitionName = competitionName;
        this.data = data; 

    }

}

/**
 * a class representing a match 
 */
export class Match
{

    constructor(id,date,home,away,competitionID,competitionName)
    {

        this.id = id;
        this.date = date;
        this.home = home;
        this.away = away;
        this.competitionID = competitionID;
        this.competitionName = competitionName;

    }

}

class InsertMatch
{

    constructor(id,competitionID,date,home,away,data)
    {

        this.id = id;
        this.competitionID = competitionID;
        this.date = date;
        this.home = home;
        this.away = away;
        this.data = data;
          

    }

}

/**
 * A class representing an edit 
 */
export class Edit 
{

    constructor(editID,structuredDataID,unstructuredDataID,isCorpus,settings,replace,replaceWith,type,order)
    {

        this.editID = editID;
        this.structuredDataID = structuredDataID;
        this.unstructuredDataID = unstructuredDataID;
        this.isCorpus = isCorpus;
        this.settings = settings;
        this.replace = replace;
        this.replaceWith = replaceWith;
        this.type = type;
        this.order = order;
        
    }

}


export class Competition
{

    constructor(id,name,countryName,countryId)
    {

        this.id = id; 
        this.name = name; 
        this.countryName = countryName;
        this.countryId = countryId

    }

}

export class ImportRequest
{

    constructor(begin,end,compId)
    {

        this.begin = begin; //optional

        this.end = end; //optional

        this.compId = compId; //mandatory 

    }

}


export class SearchRequest
{

    constructor(type,value,field)
    {

        this.type = type;
        this.value = value;
        this.field = field; 

    }

}


export class User
{

    constructor(id, username, hash, admin, regkey, token, apikey)
    {

        this.id = id; 
        this.username = username; 
        this.hash = hash;
        this.admin = admin;
        this.regkey = regkey;
        this.token = token;
        this.apikey = apikey;

    }

}