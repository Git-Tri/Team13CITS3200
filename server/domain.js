/**
 * A class representing unstructured data
 * for list you may want to leave the data field undefined
 * NOTE: this code is duplicated on client and server so is only tested on the server
 */
class UnstructuredData
{

    constructor(id,matchid,title,author,url,published,extracted,data)
    {

        this.id = id;
        this.matchid = matchid;
        this.title = title;
        this.author = author;
        this.url = url;
        this.published = published;
        this.extracted = extracted;
        this.data = data; 

    }

}

/**
 * A class representing structured data, for lists you may want to leave data undefined
 */
class StructuredData
{

    constructor(id,date,home,away,competitionID,competitionName,data)
    {

        this.id = id;
        this.date = date;
        this.home = home;
        this.away = away;
        this.competitionID = competitionID;
        this.competitionName = competitionName;
        this.data = data; 

    }

}

/**
 * a class representing a match 
 */
class Match
{

    constructor(id,date,home,away,competitionID,competitionName)
    {

        this.id = id;
        this.date = date;
        this.home = home;
        this.away = away;
        this.competitionID = competitionID;
        this.competitionName = competitionName;

    }

}

class InsertMatch
{

    constructor(id,competitionID,date,home,away,data)
    {

        this.id = id;
        this.competitionID = competitionID;
        this.date = date;
        this.home = home;
        this.away = away;
        this.data = data;
          

    }

}

/**
 * A class representing an edit 
 */
class Edit 
{

    constructor(editID,structuredDataID,unstructuredDataID,isCorpus,settings,replace,replaceWith,type)
    {

        this.editID = editID;
        this.structuredDataID = structuredDataID;
        this.unstructuredDataID = unstructuredDataID;
        this.isCorpus = isCorpus;
        this.settings = settings;
        this.replace = replace;
        this.replaceWith = replaceWith;
        this.type = type;
        
    }

}

class Competition
{

    constructor(id,name,countryName,countryId)
    {

        this.id = id; 
        this.name = name; 
        this.countryName = countryName;
        this.countryId = countryId

    }

}

class User
{

    constructor(id, username, hash, admin, regkey)
    {

        this.id = id; 
        this.username = username; 
        this.hash = hash;
        this.admin = admin;
        this.regkey = regkey;

    }

}


class ImportRequest
{

    constructor(begin,end,compId)
    {

        this.begin = begin; //optional

        this.end = end; //optional

        if(typeof(compId !== "string"))
        {

            throw new Error("Competition must exist and of type string");

        }

        this.compId = compId; //mandatory 

    }

}

class SearchRequest
{

    constructor(type,value,field)
    {

        this.type = type;
        this.value = value;
        this.field = field; 

    }

}

//export all classes 
module.exports = 
{

    UnstructuredData,StructuredData,Match,Edit,Competition,ImportRequest, InsertMatch, SearchRequest, User

}