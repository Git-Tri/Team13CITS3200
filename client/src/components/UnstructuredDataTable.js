import React, { Component } from 'react';
import { Table } from 'semantic-ui-react';
import { UnstructuredData} from "../domain";
import {MAX_TITLE_SIZE} from "../Constants";

class UnstructuredDataTable extends Component
{

    constructor(props)
    {

        super(props);

        let selectFunc;

        if(this.props.onSelect != undefined)
        {

            selectFunc = (data) => 
                {
                    
                    this.setState({activeRow: data.id})

                    this.props.onSelect(data);

                }

        }
        else
        {

            selectFunc = () => {};

        }
 

        this.state = {activeRow: -1,selectFunc: selectFunc};



    }

    render()
    {

        if(! Array.isArray( this.props.items))
        {


            throw new Error("Props.items should contain a list of unstructured data")


        }
        if(this.props.items.every((i) => i != undefined && i != null) == false)
        {

            throw new Error("props.data must be an instance of unstructured data")

        }
        //broken into each predicate for easier debugging
        if(this.props.items.every((i) => {
            let isIdValid = typeof(i.id) === "number" 
            let isPublishedValid = i.published instanceof Date 
            let isAuthorValid = typeof(i.author) === "string"
            let isTitleValid = typeof(i.title) === "string"
            return isIdValid && isPublishedValid && isAuthorValid && isTitleValid;
            }) == false)
        {
            

            
            throw new Error("every piece of data should have valid id, published, title and author");

        }

        let isSelectable = this.props.onSelect != undefined;

        


        let rows = this.props.items.map((item, index) => (

            <UnstructuredDataRow 
                key={item.id} 
                data={item} 
                onSelect={this.state.selectFunc} 
                isActive={item.id == this.state.activeRow}
            />
        ))

        return(
        <Table striped selectable={isSelectable} >        
            <Table.Header>
                <Table.Row>
                    <Table.HeaderCell>Published</Table.HeaderCell>
                    <Table.HeaderCell>Author</Table.HeaderCell>
                    <Table.HeaderCell>Title</Table.HeaderCell>
                </Table.Row>
            </Table.Header>  
            <Table.Body>
                {rows}
            </Table.Body>        
        </Table>
        );
    }
}

function UnstructuredDataRow(props)
{

    if((props.data instanceof UnstructuredData) == false)
    {

        throw Error("props.data must be an instance of unstructured data");

    }
    
    let date = props.data.published;

    let dateString = (date.getDay()+1) + "/" + (date.getMonth()+1) + "/" + date.getFullYear();

    let title = props.data.title;

    if(title.length > MAX_TITLE_SIZE)
    {

        title = title.substring(0,MAX_TITLE_SIZE-3) + "...";

    }

    let selectFunc = props.onSelect != undefined ? props.onSelect : () => {};

    return(
        <Table.Row onClick={() => selectFunc(props.data)} active={props.isActive}>
            <Table.Cell>{dateString}</Table.Cell>
            <Table.Cell>{props.data.author}</Table.Cell>
            <Table.Cell>{title}</Table.Cell>
        </Table.Row>
    );
    
}

export default UnstructuredDataTable;