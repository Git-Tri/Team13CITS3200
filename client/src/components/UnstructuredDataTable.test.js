import UnstructuredDataTable from "./UnstructuredDataTable";
import renderer from 'react-test-renderer';
import React, { Component } from 'react';
import ErrorTester from "./ErrorTester";
import {UnstructuredData} from "../domain";
import {filterConsoleError,unfilterConsoleError} from "../TestUtils";
import {shallow, mount} from "enzyme";

import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });


describe("Unstructured Data Table Tests",function()
{



    beforeAll(() =>
        {

            return new Promise(function(resolve,reject)
            {

                filterConsoleError();

                resolve();

            })

        });

    /*
    const testStructuredData = [
        new StructuredData(1,new Date(),"some team","some other team",1,"some comp","some plan",{}),
        new StructuredData(2,new Date(),"some team","some other team",1,"some comp","some plan",{})
        ];
    */
    const testUnstructuredData = [
        new UnstructuredData(1,1,"some title","some author","some url",new Date(),new Date(),"some data"),
        new UnstructuredData(2,1,"some title really really really really really long title","some author","some url",new Date(),new Date(),"some data"),
        ];

    

    test("Should throw error- when given undefined items",(done) => 
    {

    renderer.create(<ErrorTester              
             onError={(e,em) => {
                    expect(e.message).toEqual("Props.items should contain a list of unstructured data"); 
                    done();}
                    }>
                 <UnstructuredDataTable ></UnstructuredDataTable>
             </ErrorTester>
            );       

    });

    test("Should throw error when given non-array items",(done) => 
    {

        renderer.create(<ErrorTester              
            onError={(e,em) => {
                   expect(e.message).toEqual("Props.items should contain a list of unstructured data"); 
                   done();}
                   }>
                <UnstructuredDataTable items={"not an item"}></UnstructuredDataTable>
            </ErrorTester>
           );       

    });

    test("Should throw error when one item is undefined",(done) => 
    {

        let missingItemData = testUnstructuredData.slice();

        missingItemData.push(undefined);


        renderer.create(<ErrorTester              
            onError={(e,em) => {
                   expect(e.message).toEqual("props.data must be an instance of unstructured data"); 
                   done();}
                   }>
                <UnstructuredDataTable items={missingItemData}></UnstructuredDataTable>
            </ErrorTester>
           );       

    });

    test("Should throw error when item id wrong type",(done) => 
    {

        let wrongIDItemData = testUnstructuredData.slice();

        wrongIDItemData.push(new UnstructuredData("bob",1,"some title really really really really really long title","some author","some url",new Date(),new Date(),"some data"));

        renderer.create(<ErrorTester              
            onError={(e,em) => {
                   expect(e.message).toEqual("every piece of data should have valid id, published, title and author"); 
                   done();}
                   }>
                <UnstructuredDataTable items={wrongIDItemData}></UnstructuredDataTable>
            </ErrorTester>
           );       

    });

    test("Should throw error when item date wrong type",(done) => 
    {

        let wrongIDItemData = testUnstructuredData.slice();

        wrongIDItemData.push(new UnstructuredData(1,1,"some title really really really really really long title","some author","some url","some date",new Date(),"some data"));

        renderer.create(<ErrorTester              
            onError={(e,em) => {
                   expect(e.message).toEqual("every piece of data should have valid id, published, title and author"); 
                   done();}
                   }>
                <UnstructuredDataTable items={wrongIDItemData}></UnstructuredDataTable>
            </ErrorTester>
           );       

    });

    test("Should throw error when item author wrong type",(done) => 
    {

        let wrongIDItemData = testUnstructuredData.slice();

        wrongIDItemData.push(new UnstructuredData(1,1,"some title really really really really really long title",undefined,"some url",new Date(),new Date(),"some data"));

        renderer.create(<ErrorTester              
            onError={(e,em) => {
                   expect(e.message).toEqual("every piece of data should have valid id, published, title and author"); 
                   done();}
                   }>
                <UnstructuredDataTable items={wrongIDItemData}></UnstructuredDataTable>
            </ErrorTester>
           );       

    });

    test("Should throw error when item title wrong type",(done) => 
    {

        let wrongIDItemData = testUnstructuredData.slice();

        wrongIDItemData.push(new UnstructuredData(1,1,null,"some author","some url",new Date(),new Date(),"some data"));

        renderer.create(<ErrorTester              
            onError={(e,em) => {
                   expect(e.message).toEqual("every piece of data should have valid id, published, title and author"); 
                   done();}
                   }>
                <UnstructuredDataTable items={wrongIDItemData}></UnstructuredDataTable>
            </ErrorTester>
           );       

    });


    test("Should render correctly for correct input",() => 
    {        
                 
        let Component = renderer.create(<UnstructuredDataTable items={testUnstructuredData.slice()}/>)

        expect(Component.toJSON()).toMatchSnapshot();

    })

    test("Should render correctly for correct input with onSelect handler",() => 
    {        
        
        let Component = renderer.create(<UnstructuredDataTable onSelect={() => {}} items={testUnstructuredData.slice()}/>)

        expect(Component.toJSON()).toMatchSnapshot();

    })

    test("Should call callback when onselect is trigerred",(done) => 
    {

        let wrapper = mount(<UnstructuredDataTable onSelect={() => done()} items={testUnstructuredData.slice()}/>)

        wrapper.state().selectFunc(testUnstructuredData[0]);


    })

    test("Should render correctly with selected row",() => 
    {

        let wrapper = shallow(<UnstructuredDataTable onSelect={() => {}} items={testUnstructuredData.slice()}/>)

        wrapper.state().selectFunc(testUnstructuredData[0]);

        expect(wrapper).toMatchSnapshot();




    })

    afterAll(() =>
    {

        return new Promise(function(resolve,reject)
        {

            unfilterConsoleError();

            resolve();

        })

    });

})

