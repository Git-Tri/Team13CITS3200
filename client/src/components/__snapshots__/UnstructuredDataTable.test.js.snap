// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unstructured Data Table Tests Should render correctly for correct input 1`] = `
<table
  className="ui striped table"
>
  <thead
    className=""
  >
    <tr
      className=""
    >
      <th
        className=""
      >
        Published
      </th>
      <th
        className=""
      >
        Author
      </th>
      <th
        className=""
      >
        Title
      </th>
    </tr>
  </thead>
  <tbody
    className=""
  >
    <tr
      className=""
      onClick={[Function]}
    >
      <td
        className=""
      >
        7/4/1991
      </td>
      <td
        className=""
      >
        some author
      </td>
      <td
        className=""
      >
        some title
      </td>
    </tr>
    <tr
      className=""
      onClick={[Function]}
    >
      <td
        className=""
      >
        7/4/1991
      </td>
      <td
        className=""
      >
        some author
      </td>
      <td
        className=""
      >
        some title r...
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`Unstructured Data Table Tests Should render correctly for correct input with onSelect handler 1`] = `
<table
  className="ui selectable striped table"
>
  <thead
    className=""
  >
    <tr
      className=""
    >
      <th
        className=""
      >
        Published
      </th>
      <th
        className=""
      >
        Author
      </th>
      <th
        className=""
      >
        Title
      </th>
    </tr>
  </thead>
  <tbody
    className=""
  >
    <tr
      className=""
      onClick={[Function]}
    >
      <td
        className=""
      >
        7/4/1991
      </td>
      <td
        className=""
      >
        some author
      </td>
      <td
        className=""
      >
        some title
      </td>
    </tr>
    <tr
      className=""
      onClick={[Function]}
    >
      <td
        className=""
      >
        7/4/1991
      </td>
      <td
        className=""
      >
        some author
      </td>
      <td
        className=""
      >
        some title r...
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`Unstructured Data Table Tests Should render correctly for correct with no data 1`] = `
<table
  className="ui selectable striped table"
>
  <thead
    className=""
  >
    <tr
      className=""
    >
      <th
        className=""
      >
        Published
      </th>
      <th
        className=""
      >
        Author
      </th>
      <th
        className=""
      >
        Title
      </th>
    </tr>
  </thead>
  <tbody
    className=""
  />
</table>
`;

exports[`Unstructured Data Table Tests Should render correctly with selected row 1`] = `
<Table
  as="table"
  selectable={true}
  striped={true}
>
  <TableHeader
    as="thead"
  >
    <TableRow
      as="tr"
      cellAs="td"
    >
      <TableHeaderCell
        as="th"
      >
        Published
      </TableHeaderCell>
      <TableHeaderCell
        as="th"
      >
        Author
      </TableHeaderCell>
      <TableHeaderCell
        as="th"
      >
        Title
      </TableHeaderCell>
    </TableRow>
  </TableHeader>
  <TableBody
    as="tbody"
  >
    <UnstructuredDataRow
      data={
        UnstructuredData {
          "author": "some author",
          "data": "some data",
          "extracted": 1991-04-20T00:00:00.000Z,
          "id": 1,
          "matchid": 1,
          "published": 1991-04-20T00:00:00.000Z,
          "title": "some title",
          "url": "some url",
        }
      }
      isActive={true}
      key="1"
      onSelect={[Function]}
    />
    <UnstructuredDataRow
      data={
        UnstructuredData {
          "author": "some author",
          "data": "some data",
          "extracted": 1991-04-20T00:00:00.000Z,
          "id": 2,
          "matchid": 1,
          "published": 1991-04-20T00:00:00.000Z,
          "title": "some title really really really really really long title",
          "url": "some url",
        }
      }
      isActive={false}
      key="2"
      onSelect={[Function]}
    />
  </TableBody>
</Table>
`;
